File: podklady/6_Others/2_github_copilot_sdk/typescript/src/2_multi_agent/1_collaboration_pattern.ts
/**
 * Multi-agent — Collaboration pattern.
 *
 * Concept
 * -------
 * A flat chain of equal peers. Each agent sees the previous agent's output,
 * either resolves the task (`resolved: true`) or hands off to the next one
 * in the predefined sequence. No supervisor.
 *
 * Run:
 *   npx tsx src/2_multi_agent/1_collaboration_pattern.ts
 */

import { CopilotClient, approveAll } from "@github/copilot-sdk";

interface Role {
    name: string;
    prompt: string;
}

const PIPELINE: Role[] = [
    {
        name: "requirements-analyst",
        prompt:
            "You are a requirements analyst. Translate the user's request into a " +
            "short, concrete list of functional requirements. No code.",
    },
    {
        name: "architect",
        prompt:
            "You are a software architect. Given the requirements, propose a tiny " +
            "REST API design: endpoints, request/response shape, data model.",
    },
    {
        name: "developer",
        prompt:
            "You are a TypeScript developer. Given the architecture, sketch an Express " +
            "implementation in <80 lines. Include 1 example request.",
    },
];

const REPLY_FORMAT = `
After your work, append exactly one JSON block on its own line:

\`\`\`json
{"resolved": true|false, "next_input": "<text for the next agent>"}
\`\`\`

Set \`resolved: true\` only if no further refinement is needed.
`.trim();

function parseDecision(text: string): { resolved: boolean; nextInput: string } {
    const m = text.match(/```json\s*(\{[\s\S]*?\})\s*```/);
    if (!m) return { resolved: false, nextInput: text };
    try {
        const obj = JSON.parse(m[1]) as { resolved?: boolean; next_input?: string };
        return { resolved: !!obj.resolved, nextInput: obj.next_input ?? text };
    } catch {
        return { resolved: false, nextInput: text };
    }
}

async function runRole(
    client: CopilotClient,
    role: Role,
    priorOutput: string,
): Promise<{ resolved: boolean; nextInput: string }> {
    const session = await client.createSession({
        onPermissionRequest: approveAll,
        systemMessage: { mode: "replace", content: `${role.prompt}\n\n${REPLY_FORMAT}` },
    });
    try {
        const reply = await session.sendAndWait({ prompt: priorOutput }, 300_000);
        const text = (reply?.data as { content: string })?.content ?? "";
        const decision = parseDecision(text);
        console.log(`\n=== ${role.name} ===`);
        console.log(text);
        return decision;
    } finally {
        await session.disconnect();
    }
}

async function main(): Promise<void> {
    const initial =
        "Design a small REST API for a personal todo list with create/list/" +
        "update/delete operations. Keep it minimal.";

    const client = new CopilotClient();
    try {
        let currentInput = initial;
        for (const role of PIPELINE) {
            const { resolved, nextInput } = await runRole(client, role, currentInput);
            currentInput = nextInput;
            if (resolved) {
                console.log(`\n[stopping early — ${role.name} marked resolved]`);
                break;
            }
        }
    } finally {
        await client.stop();
    }
}

main().catch(console.error);

File: podklady/6_Others/2_github_copilot_sdk/python/2_multi_agent/1_collaboration_pattern.py
"""
Multi-agent — Collaboration pattern.

Concept
-------
A flat chain of equal peers. Each agent sees the previous agent's output,
either resolves the task (`resolved: true`) or hands off to the next one in
the predefined sequence. No supervisor.

Scenario
--------
Build a tiny REST API design:
    requirements-analyst → architect → developer

Each role gets a fresh `CopilotSession` with a tailored system prompt.

Run:
    python 2_multi_agent/1_collaboration_pattern.py
"""

import asyncio
import json
import re
from dataclasses import dataclass

from copilot import CopilotClient
from copilot.session import PermissionHandler


@dataclass
class Role:
    name: str
    prompt: str


PIPELINE: list[Role] = [
    Role(
        "requirements-analyst",
        "You are a requirements analyst. Translate the user's request into a "
        "short, concrete list of functional requirements. No code.",
    ),
    Role(
        "architect",
        "You are a software architect. Given the requirements, propose a tiny "
        "REST API design: endpoints, request/response shape, data model.",
    ),
    Role(
        "developer",
        "You are a Python developer. Given the architecture, sketch a FastAPI "
        "implementation in <80 lines. Include 1 example request.",
    ),
]


REPLY_FORMAT = """
After your work, append exactly one JSON block on its own line:

```json
{"resolved": true|false, "next_input": "<text for the next agent>"}
```

Set `resolved: true` only if no further refinement is needed.
""".strip()


def parse_decision(text: str) -> tuple[bool, str]:
    m = re.search(r"```json\s*(\{.*?\})\s*```", text, re.DOTALL)
    if not m:
        return False, text
    try:
        obj = json.loads(m.group(1))
        return bool(obj.get("resolved")), str(obj.get("next_input", text))
    except json.JSONDecodeError:
        return False, text


async def run_role(client: CopilotClient, role: Role, prior_output: str) -> tuple[bool, str]:
    session = await client.create_session(
        on_permission_request=PermissionHandler.approve_all,
        system_message={"mode": "replace", "content": f"{role.prompt}\n\n{REPLY_FORMAT}"},
    )
    try:
        reply = await session.send_and_wait(prior_output, timeout=300.0)
        text = reply.data.content if reply else ""
        resolved, next_input = parse_decision(text)
        print(f"\n=== {role.name} ===")
        print(text)
        return resolved, next_input
    finally:
        await session.disconnect()


async def main() -> None:
    initial = (
        "Design a small REST API for a personal todo list with create/list/"
        "update/delete operations. Keep it minimal."
    )

    client = CopilotClient()
    await client.start()
    try:
        current_input = initial
        for role in PIPELINE:
            resolved, current_input = await run_role(client, role, current_input)
            if resolved:
                print(f"\n[stopping early — {role.name} marked resolved]")
                break
    finally:
        await client.stop()


if __name__ == "__main__":
    asyncio.run(main())
